File: packages/store/src/atomic.ts
// ─── @mirk/store/atomic ────────────────────────────────────────────────────
// Optional, declarative atomic mutation capability. This module is deliberately
// dependency-free (including the digest implementation) so it remains safe to
// import from the root package and browser-facing ports.

import type { AsyncStore, SyncStore } from "./types.js";
// `order.ts` and `canonical.ts` are dependency-free, so this keeps the module
// dependency-free.
import { compareCodePoints } from "./order.js";
import { canonicalJson, isPlainObject, sha256Hex } from "./canonical.js";

/** @deprecated Spelled `compareCodePoints` in `./order.js`; kept for the
 *  published `@mirk/store/atomic` surface. */
export { compareCodePoints as compareCodePoint } from "./order.js";

/** Canonical JSON used by request digests and bounded payload checks.
 *  Implemented in `./canonical.js`; re-exported here because
 *  `@mirk/store/atomic` and the package root are its published homes. */
export { canonicalJson } from "./canonical.js";

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | readonly JsonValue[]
  | { readonly [key: string]: JsonValue };

export type JsonObject = { readonly [key: string]: JsonValue };

export type StoreTarget =
  | { kind: "key"; key: string }
  | { kind: "record"; collection: string; id: string };

declare const storeVersionBrand: unique symbol;
export type StoreVersion = string & { readonly [storeVersionBrand]: true };

export interface VersionedStoreValue<T> {
  value: T;
  version: StoreVersion;
}

export interface SyncVersionedReadStore {
  getVersioned<T>(target: StoreTarget): VersionedStoreValue<T> | null;
}

export interface AsyncVersionedReadStore {
  getVersioned<T>(target: StoreTarget): Promise<VersionedStoreValue<T> | null>;
}

export type StoreCondition =
  | { target: StoreTarget; expected: "missing" }
  | { target: StoreTarget; expected: "present" }
  | { target: StoreTarget; expected: "version"; version: StoreVersion };

export type AtomicStoreOperation =
  | { op: "set"; key: string; value: JsonValue }
  | { op: "delete"; key: string }
  | { op: "put"; collection: string; item: { id: string } & JsonObject }
  | { op: "remove"; collection: string; id: string };

export interface AtomicIdempotency {
  key: string;
  outcome?: JsonValue;
}

export interface AtomicMutationRequest {
  conditions?: readonly StoreCondition[];
  operations: readonly AtomicStoreOperation[];
  idempotency?: AtomicIdempotency;
}

export type AtomicMutationResult =
  | {
      status: "applied" | "replayed";
      requestDigest: string;
      versions: readonly {
        target: StoreTarget;
        version: StoreVersion | null;
      }[];
      outcome?: JsonValue;
    }
  | {
      status: "conflict";
      condition: StoreCondition;
      observed: "missing" | "present" | StoreVersion;
    }
  | {
      status: "idempotency-conflict";
      key: string;
      expectedRequestDigest: string;
      receivedRequestDigest: string;
    };

export type AtomicCompletedMutationResult = Extract<
  AtomicMutationResult,
  { status: "applied" | "replayed" }
>;

export type AtomicMutationRejectionCode =
  | "invalid-request"
  | "unsupported-operation"
  | "condition-limit-exceeded"
  | "operation-limit-exceeded"
  | "request-size-exceeded"
  | "outcome-size-exceeded";

export class AtomicMutationRejectedError extends Error {
  readonly name = "AtomicMutationRejectedError";
  constructor(readonly code: AtomicMutationRejectionCode, message: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AtomicMutationBackendError extends Error {
  readonly name = "AtomicMutationBackendError";
  constructor(
    readonly code: "unavailable" | "serialization-failure",
    readonly retryable: boolean,
    message: string
  ) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AtomicMutationIndeterminateError extends Error {
  readonly name = "AtomicMutationIndeterminateError";
  constructor(
    readonly requestDigest: string,
    readonly idempotencyKey: string | undefined,
    readonly recovery: "retry-with-same-key" | "manual-reconciliation",
    message = "Atomic mutation outcome is indeterminate."
  ) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/** The request bounds one store applies before its atomic decision point.
 *
 * These are a wire-contract guard, so their right value depends on how far the
 * request travels. A remote adapter serializes the batch over a network and
 * wants the conservative default; an embedded backend never leaves the process
 * and can afford far more. Each store publishes what it enforces.
 */
export interface AtomicMutationLimits {
  readonly maxOperations: number;
  readonly maxConditions: number;
  readonly maxRequestBytes: number;
}

/** The portable defaults: what every store enforced before limits were
 *  configurable, and what a remote or unknown transport should keep. */
export const DEFAULT_ATOMIC_LIMITS: AtomicMutationLimits = Object.freeze({
  maxOperations: 128,
  maxConditions: 128,
  maxRequestBytes: 1024 * 1024,
});

/** In-process backends (the reference and the SQLite adapter) — the request is
 *  never serialized onto a wire and the batch is one local transaction. */
export const IN_PROCESS_ATOMIC_LIMITS: AtomicMutationLimits = Object.freeze({
  maxOperations: 4096,
  maxConditions: 1024,
  maxRequestBytes: 16 * 1024 * 1024,
});

function limitField(
  overrides: Partial<AtomicMutationLimits> | undefined,
  base: AtomicMutationLimits,
  field: keyof AtomicMutationLimits
): number {
  const value = overrides?.[field];
  if (value === undefined) return base[field];
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new Error(
      `${field} must be a positive safe integer; got ${String(value)}.`
    );
  }
  return value;
}

/** Merge caller overrides onto a backend's own defaults. */
export function resolveAtomicLimits(
  overrides?: Partial<AtomicMutationLimits>,
  base: AtomicMutationLimits = DEFAULT_ATOMIC_LIMITS
): AtomicMutationLimits {
  return Object.freeze({
    maxOperations: limitField(overrides, base, "maxOperations"),
    maxConditions: limitField(overrides, base, "maxConditions"),
    maxRequestBytes: limitField(overrides, base, "maxRequestBytes"),
  });
}

export interface SyncAtomicMutationStore extends SyncVersionedReadStore {
  /** The request bounds this store enforces. */
  readonly atomicLimits: AtomicMutationLimits;
  mutateAtomically(request: AtomicMutationRequest): AtomicMutationResult;
}

export interface AsyncAtomicMutationStore extends AsyncVersionedReadStore {
  /** The request bounds this store enforces. */
  readonly atomicLimits: AtomicMutationLimits;
  mutateAtomically(
    request: AtomicMutationRequest
  ): Promise<AtomicMutationResult>;
}

function hasAtomicLimits(candidate: {
  atomicLimits?: AtomicMutationLimits;
}): boolean {
  const limits = candidate.atomicLimits;
  return (
    typeof limits === "object" &&
    limits !== null &&
    typeof limits.maxOperations === "number" &&
    typeof limits.maxConditions === "number" &&
    typeof limits.maxRequestBytes === "number"
  );
}

export function supportsAtomicMutation(
  store: SyncStore
): store is SyncStore & SyncAtomicMutationStore {
  const candidate = store as Partial<SyncAtomicMutationStore>;
  return (
    typeof candidate.getVersioned === "function" &&
    typeof candidate.mutateAtomically === "function" &&
    hasAtomicLimits(candidate)
  );
}

export function supportsAsyncAtomicMutation(
  store: AsyncStore
): store is AsyncStore & AsyncAtomicMutationStore {
  const candidate = store as Partial<AsyncAtomicMutationStore>;
  return (
    typeof candidate.getVersioned === "function" &&
    typeof candidate.mutateAtomically === "function" &&
    hasAtomicLimits(candidate)
  );
}

/** @deprecated Read `store.atomicLimits.maxConditions`, or
 *  `DEFAULT_ATOMIC_LIMITS.maxConditions` for the portable default. */
export const MAX_ATOMIC_CONDITIONS = DEFAULT_ATOMIC_LIMITS.maxConditions;
/** @deprecated Read `store.atomicLimits.maxOperations`, or
 *  `DEFAULT_ATOMIC_LIMITS.maxOperations` for the portable default. */
export const MAX_ATOMIC_OPERATIONS = DEFAULT_ATOMIC_LIMITS.maxOperations;
/** @deprecated Read `store.atomicLimits.maxRequestBytes`, or
 *  `DEFAULT_ATOMIC_LIMITS.maxRequestBytes` for the portable default. */
export const MAX_ATOMIC_REQUEST_BYTES = DEFAULT_ATOMIC_LIMITS.maxRequestBytes;
/** A fixed cap, not configurable: an idempotency outcome is persisted under its
 *  key forever, so no backend may accept a larger one. */
export const MAX_ATOMIC_OUTCOME_BYTES = 64 * 1024;

const REQUEST_SCHEMA = "mirk-atomic-request/v1";
const encoder = new TextEncoder();

export function compareTargets(a: StoreTarget, b: StoreTarget): number {
  if (a.kind !== b.kind) return a.kind === "key" ? -1 : 1;
  if (a.kind === "key" && b.kind === "key")
    return compareCodePoints(a.key, b.key);
  if (a.kind === "record" && b.kind === "record") {
    return (
      compareCodePoints(a.collection, b.collection) ||
      compareCodePoints(a.id, b.id)
    );
  }
  return 0;
}

export function targetKey(target: StoreTarget): string {
  // Length-prefix each component. Raw separators are not sufficient because
  // callers may legitimately use those code points in keys, collection names,
  // or ids (for example `collection: "a\0b", id: "c"` versus `"a", "b\0c"`).
  const part = (value: string): string => `${value.length}:${value}`;
  return target.kind === "key"
    ? `k:${part(target.key)}`
    : `r:${part(target.collection)}:${part(target.id)}`;
}

export function cloneJson<T extends JsonValue>(value: T): T {
  // The validator below guarantees JSON values. Parsing the canonical encoding
  // also gives in-memory receipts the same value-copy behavior as SQLite.
  return JSON.parse(canonicalJson(value)) as T;
}

interface ValidatedRequest {
  conditions: StoreCondition[];
  operations: AtomicStoreOperation[];
  idempotency?: AtomicIdempotency;
  requestDigest: string;
  outcome?: JsonValue;
}

function invalid(message: string): never {
  throw new AtomicMutationRejectedError("invalid-request", message);
}

function targetFromUnknown(value: unknown): StoreTarget {
  if (
    !isPlainObject(value) ||
    (value.kind !== "key" && value.kind !== "record")
  )
    invalid("invalid store target");
  if (value.kind === "key") {
    if (typeof value.key !== "string")
      invalid("key target requires a string key");
    return { kind: "key", key: value.key };
  }
  if (
    typeof value.collection !== "string" ||
    value.collection.length === 0 ||
    typeof value.id !== "string"
  ) {
    invalid("record target requires string collection and id");
  }
  return { kind: "record", collection: value.collection, id: value.id };
}

function jsonSafe(value: unknown, label: string): asserts value is JsonValue {
  try {
    canonicalJson(value);
  } catch (error) {
    invalid(
      `${label} is not JSON-safe: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function operationTarget(operation: AtomicStoreOperation): StoreTarget {
  switch (operation.op) {
    case "set":
    case "delete":
      return { kind: "key", key: operation.key };
    case "put":
      return {
        kind: "record",
        collection: operation.collection,
        id: operation.item.id,
      };
    case "remove":
      return {
        kind: "record",
        collection: operation.collection,
        id: operation.id,
      };
    default:
      throw new AtomicMutationRejectedError(
        "unsupported-operation",
        "unsupported atomic operation"
      );
  }
}

function normalizeOperation(value: unknown): AtomicStoreOperation {
  if (!isPlainObject(value) || typeof value.op !== "string")
    invalid("invalid atomic operation");
  switch (value.op) {
    case "set":
      if (typeof value.key !== "string") invalid("set requires a string key");
      jsonSafe(value.value, "set value");
      return { op: "set", key: value.key, value: cloneJson(value.value) };
    case "delete":
      if (typeof value.key !== "string")
        invalid("delete requires a string key");
      return { op: "delete", key: value.key };
    case "put":
      if (
        typeof value.collection !== "string" ||
        value.collection.length === 0 ||
        !isPlainObject(value.item) ||
        typeof value.item.id !== "string"
      ) {
        invalid("put requires a collection and an item with a string id");
      }
      jsonSafe(value.item, "put item");
      return {
        op: "put",
        collection: value.collection,
        item: cloneJson(value.item) as { id: string } & JsonObject,
      };
    case "remove":
      if (
        typeof value.collection !== "string" ||
        value.collection.length === 0 ||
        typeof value.id !== "string"
      )
        invalid("remove requires string collection and id");
      return { op: "remove", collection: value.collection, id: value.id };
    default:
      throw new AtomicMutationRejectedError(
        "unsupported-operation",
        `unsupported atomic operation: ${value.op}`
      );
  }
}

function normalizeCondition(value: unknown): StoreCondition {
  if (!isPlainObject(value)) invalid("invalid atomic condition");
  const target = targetFromUnknown(value.target);
  if (value.expected === "missing" || value.expected === "present")
    return { target, expected: value.expected };
  if (value.expected === "version" && typeof value.version === "string") {
    return {
      target,
      expected: "version",
      version: value.version as StoreVersion,
    };
  }
  invalid("invalid atomic condition expectation");
}

/** Validate and canonicalize a request before any backend decision point. */
export function validateAtomicRequest(
  request: AtomicMutationRequest,
  limits: AtomicMutationLimits = DEFAULT_ATOMIC_LIMITS
): ValidatedRequest {
  if (!isPlainObject(request)) invalid("request must be a plain object");
  if (!Array.isArray(request.operations) || request.operations.length === 0)
    invalid("operations must be a non-empty array");
  if (request.operations.length > limits.maxOperations) {
    throw new AtomicMutationRejectedError(
      "operation-limit-exceeded",
      `request has ${request.operations.length} operations; this store's maxOperations is ${limits.maxOperations}`
    );
  }
  if (request.conditions !== undefined && !Array.isArray(request.conditions))
    invalid("conditions must be an array");
  const conditionCount = request.conditions?.length ?? 0;
  if (conditionCount > limits.maxConditions) {
    throw new AtomicMutationRejectedError(
      "condition-limit-exceeded",
      `request has ${conditionCount} conditions; this store's maxConditions is ${limits.maxConditions}`
    );
  }

  const conditions: StoreCondition[] = [];
  for (let index = 0; index < (request.conditions?.length ?? 0); index += 1) {
    if (!(index in (request.conditions ?? [])))
      invalid("conditions must not be sparse");
    conditions.push(normalizeCondition(request.conditions![index]));
  }
  const conditionKeys = new Set<string>();
  for (const condition of conditions) {
    const key = targetKey(condition.target);
    if (conditionKeys.has(key))
      invalid("repeated conditions for one target are not supported");
    conditionKeys.add(key);
  }
  conditions.sort((a, b) => compareTargets(a.target, b.target));

  const operations: AtomicStoreOperation[] = [];
  for (let index = 0; index < request.operations.length; index += 1) {
    if (!(index in request.operations))
      invalid("operations must not be sparse");
    operations.push(normalizeOperation(request.operations[index]));
  }
  const operationKeys = new Set<string>();
  for (const operation of operations) {
    const key = targetKey(operationTarget(operation));
    if (operationKeys.has(key))
      invalid("repeated operation targets are not supported");
    operationKeys.add(key);
  }

  let idempotency: AtomicIdempotency | undefined;
  let outcome: JsonValue | undefined;
  if (request.idempotency !== undefined) {
    if (
      !isPlainObject(request.idempotency) ||
      typeof request.idempotency.key !== "string"
    )
      invalid("idempotency requires a string key");
    idempotency = { key: request.idempotency.key };
    if ("outcome" in request.idempotency) {
      jsonSafe(request.idempotency.outcome, "idempotency outcome");
      outcome = cloneJson(request.idempotency.outcome);
      const outcomeBytes = encoder.encode(canonicalJson(outcome)).byteLength;
      if (outcomeBytes > MAX_ATOMIC_OUTCOME_BYTES) {
        throw new AtomicMutationRejectedError(
          "outcome-size-exceeded",
          `outcome is ${outcomeBytes} bytes; the fixed outcome cap is ${MAX_ATOMIC_OUTCOME_BYTES} bytes`
        );
      }
      idempotency.outcome = outcome;
    }
  }

  const digestInput: Record<string, unknown> = {
    schema: REQUEST_SCHEMA,
    conditions,
    operations,
  };
  if (outcome !== undefined) digestInput.outcome = outcome;
  const encoded = canonicalJson(digestInput);
  // The idempotency key is deliberately excluded from the digest (the same
  // mutation has the same request identity under any key), but it is still
  // part of the bounded wire request and must not provide an unbounded escape
  // from MAX_ATOMIC_REQUEST_BYTES.
  const requestEncoding =
    idempotency === undefined
      ? encoded
      : canonicalJson({
          ...digestInput,
          idempotency: { key: idempotency.key },
        });
  const requestBytes = encoder.encode(requestEncoding).byteLength;
  if (requestBytes > limits.maxRequestBytes) {
    throw new AtomicMutationRejectedError(
      "request-size-exceeded",
      `request is ${requestBytes} bytes; this store's maxRequestBytes is ${limits.maxRequestBytes}`
    );
  }
  return {
    conditions,
    operations,
    idempotency,
    outcome,
    requestDigest: sha256Hex(encoded),
  };
}

File: python/store/src/mirk/store/atomic.py
"""Declarative atomic mutation: request validation, digests, and result shapes.

Port of ``packages/store/src/atomic.ts``. Validation order and every rejection
message are copied from that file, because the request digest and the errors are
part of the cross-language contract: the same request must produce the same
``requestDigest`` under Node and under Python, and a caller that switches
languages must read the same message.

Results are plain dicts with the TypeScript key spelling (``status``,
``requestDigest``, ``versions``, ``outcome``), so a result serialized by one
language deserializes into the other unchanged.
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from functools import cmp_to_key
from typing import Any, Literal, cast

from .canonical import canonical_json, sha256_hex

__all__ = [
    "DEFAULT_ATOMIC_LIMITS",
    "IN_PROCESS_ATOMIC_LIMITS",
    "MAX_ATOMIC_OUTCOME_BYTES",
    "REQUEST_SCHEMA",
    "UNDEFINED",
    "AtomicMutationBackendError",
    "AtomicMutationIndeterminateError",
    "AtomicMutationLimits",
    "AtomicMutationRejectedError",
    "StoreTarget",
    "ValidatedRequest",
    "clone_condition",
    "clone_json",
    "clone_target",
    "compare_targets",
    "condition_matches",
    "operation_target",
    "resolve_atomic_limits",
    "target_key",
    "validate_atomic_request",
]

REQUEST_SCHEMA = "mirk-atomic-request/v1"

#: A fixed cap, not configurable: an idempotency outcome is persisted under its
#: key forever, so no backend may accept a larger one.
MAX_ATOMIC_OUTCOME_BYTES = 64 * 1024

StoreTarget = dict[str, Any]
"""``{"kind": "key", "key": str}`` or ``{"kind": "record", "collection", "id"}``."""

AtomicMutationRejectionCode = Literal[
    "invalid-request",
    "unsupported-operation",
    "condition-limit-exceeded",
    "operation-limit-exceeded",
    "request-size-exceeded",
    "outcome-size-exceeded",
]


class _Undefined:
    """Stands in for JavaScript ``undefined``, which Python's ``None`` is not.

    ``{"outcome": None}`` is a request carrying a JSON null outcome; a request
    with no ``outcome`` key carries none at all. The two produce different
    digests, so the absent case needs its own value.
    """

    __slots__ = ()

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return "<undefined>"


UNDEFINED = _Undefined()


class AtomicMutationRejectedError(Exception):
    """The request never reached a backend decision point."""

    def __init__(self, code: AtomicMutationRejectionCode, message: str) -> None:
        super().__init__(message)
        self.code: AtomicMutationRejectionCode = code
        self.message = message


class AtomicMutationBackendError(Exception):
    """The backend could not decide the request; ``retryable`` says whether to retry."""

    def __init__(
        self,
        code: Literal["unavailable", "serialization-failure"],
        retryable: bool,
        message: str,
    ) -> None:
        super().__init__(message)
        self.code: Literal["unavailable", "serialization-failure"] = code
        self.retryable = retryable
        self.message = message


class AtomicMutationIndeterminateError(Exception):
    """The request may or may not have applied."""

    def __init__(
        self,
        request_digest: str,
        idempotency_key: str | None,
        recovery: Literal["retry-with-same-key", "manual-reconciliation"],
        message: str = "Atomic mutation outcome is indeterminate.",
    ) -> None:
        super().__init__(message)
        self.request_digest = request_digest
        self.idempotency_key = idempotency_key
        self.recovery: Literal["retry-with-same-key", "manual-reconciliation"] = recovery
        self.message = message


@dataclass(frozen=True, slots=True)
class AtomicMutationLimits:
    """The request bounds one store applies before its atomic decision point.

    Field names keep the TypeScript spelling so a corpus or wire ``atomicLimits``
    object maps straight in.
    """

    maxOperations: int
    maxConditions: int
    maxRequestBytes: int


#: What every store enforced before limits were configurable, and what a remote
#: or unknown transport should keep.
DEFAULT_ATOMIC_LIMITS = AtomicMutationLimits(
    maxOperations=128,
    maxConditions=128,
    maxRequestBytes=1024 * 1024,
)

#: In-process backends: the request is never serialized onto a wire and the
#: batch is one local transaction.
IN_PROCESS_ATOMIC_LIMITS = AtomicMutationLimits(
    maxOperations=4096,
    maxConditions=1024,
    maxRequestBytes=16 * 1024 * 1024,
)

_LIMIT_FIELDS = ("maxOperations", "maxConditions", "maxRequestBytes")


def _js_string(value: Any) -> str:
    """``String(value)`` for the values that reach a limit-rejection message."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return value
    if isinstance(value, float):
        if math.isnan(value):
            return "NaN"
        if math.isinf(value):
            return "Infinity" if value > 0 else "-Infinity"
    if isinstance(value, int | float):
        return canonical_json(value)
    return str(value)


def _is_safe_integer(value: Any) -> bool:
    """``Number.isSafeInteger``: a number, integral, within 2^53 - 1."""
    if isinstance(value, bool) or not isinstance(value, int | float):
        return False
    if isinstance(value, float) and (
        math.isnan(value) or math.isinf(value) or not value.is_integer()
    ):
        return False
    return abs(int(value)) <= 9007199254740991


def _limit_field(source: dict[str, Any], base: AtomicMutationLimits, name: str) -> int:
    if name not in source:
        return int(getattr(base, name))
    value = source[name]
    if not _is_safe_integer(value) or value < 1:
        raise ValueError(f"{name} must be a positive safe integer; got {_js_string(value)}.")
    return int(value)


def resolve_atomic_limits(
    overrides: AtomicMutationLimits | dict[str, Any] | None = None,
    base: AtomicMutationLimits = DEFAULT_ATOMIC_LIMITS,
) -> AtomicMutationLimits:
    """Merge caller overrides onto a backend's own defaults."""
    if isinstance(overrides, AtomicMutationLimits):
        source: dict[str, Any] = {name: getattr(overrides, name) for name in _LIMIT_FIELDS}
    else:
        source = dict(overrides) if overrides else {}
    return AtomicMutationLimits(
        maxOperations=_limit_field(source, base, "maxOperations"),
        maxConditions=_limit_field(source, base, "maxConditions"),
        maxRequestBytes=_limit_field(source, base, "maxRequestBytes"),
    )


# ── Targets ─────────────────────────────────────────────────────────────────


def _utf16_length(value: str) -> int:
    """``String.prototype.length``: UTF-16 code units, not code points.

    The prefix has to count what TypeScript counts, or an astral-plane key would
    be length-prefixed differently in the two languages and the duplicate-target
    check could disagree.
    """
    return len(value.encode("utf-16-le")) // 2


def target_key(target: StoreTarget) -> str:
    """A collision-free string identity for a target.

    Each component is length-prefixed. Raw separators are not sufficient because
    callers may legitimately use those code points in keys, collection names, or
    ids (``collection: "a\\0b", id: "c"`` versus ``"a", "b\\0c"``).
    """

    def part(value: str) -> str:
        return f"{_utf16_length(value)}:{value}"

    if target["kind"] == "key":
        return f"k:{part(target['key'])}"
    return f"r:{part(target['collection'])}:{part(target['id'])}"


def _compare_code_points(a: str, b: str) -> int:
    if a == b:
        return 0
    return -1 if a < b else 1


def compare_targets(a: StoreTarget, b: StoreTarget) -> int:
    """Key targets before record targets, then code point order within a kind."""
    if a["kind"] != b["kind"]:
        return -1 if a["kind"] == "key" else 1
    if a["kind"] == "key":
        return _compare_code_points(a["key"], b["key"])
    return _compare_code_points(a["collection"], b["collection"]) or _compare_code_points(
        a["id"], b["id"]
    )


def operation_target(operation: dict[str, Any]) -> StoreTarget:
    """The single target one operation writes."""
    op = operation.get("op")
    if op in ("set", "delete"):
        return {"kind": "key", "key": operation["key"]}
    if op == "put":
        return {
            "kind": "record",
            "collection": operation["collection"],
            "id": operation["item"]["id"],
        }
    if op == "remove":
        return {"kind": "record", "collection": operation["collection"], "id": operation["id"]}
    raise AtomicMutationRejectedError("unsupported-operation", "unsupported atomic operation")


def clone_target(target: StoreTarget) -> StoreTarget:
    if target["kind"] == "key":
        return {"kind": "key", "key": target["key"]}
    return {"kind": "record", "collection": target["collection"], "id": target["id"]}


def clone_condition(condition: dict[str, Any]) -> dict[str, Any]:
    if condition["expected"] == "version":
        return {
            "target": clone_target(condition["target"]),
            "expected": "version",
            "version": condition["version"],
        }
    return {"target": clone_target(condition["target"]), "expected": condition["expected"]}


def clone_json(value: Any) -> Any:
    """Deep copy through the canonical encoding, exactly as TypeScript does.

    Parsing the canonical encoding also gives in-memory receipts the same
    value-copy behavior as SQLite: an integral float comes back as an int.
    """
    return json.loads(canonical_json(value))


def condition_matches(condition: dict[str, Any], observed: dict[str, Any] | None) -> bool:
    expected = condition["expected"]
    if expected == "missing":
        return observed is None
    if expected == "present":
        return observed is not None
    return observed is not None and observed["version"] == condition["version"]


# ── Validation ──────────────────────────────────────────────────────────────


def _invalid(message: str) -> Any:
    raise AtomicMutationRejectedError("invalid-request", message)


def _is_plain_object(value: Any) -> bool:
    return isinstance(value, dict)


def _target_from_unknown(value: Any) -> StoreTarget:
    if not _is_plain_object(value) or value.get("kind") not in ("key", "record"):
        _invalid("invalid store target")
    if value["kind"] == "key":
        if not isinstance(value.get("key"), str):
            _invalid("key target requires a string key")
        return {"kind": "key", "key": value["key"]}
    collection = value.get("collection")
    if (
        not isinstance(collection, str)
        or len(collection) == 0
        or not isinstance(value.get("id"), str)
    ):
        _invalid("record target requires string collection and id")
    return {"kind": "record", "collection": value["collection"], "id": value["id"]}


def _json_safe(value: Any, label: str) -> None:
    if isinstance(value, _Undefined):
        _invalid(f"{label} is not JSON-safe: value is not JSON-safe")
    try:
        canonical_json(value)
    except Exception as error:
        _invalid(f"{label} is not JSON-safe: {error}")


def _normalize_operation(value: Any) -> dict[str, Any]:
    if not _is_plain_object(value) or not isinstance(value.get("op"), str):
        _invalid("invalid atomic operation")
    op = value["op"]
    if op == "set":
        if not isinstance(value.get("key"), str):
            _invalid("set requires a string key")
        payload = value.get("value", UNDEFINED)
        _json_safe(payload, "set value")
        return {"op": "set", "key": value["key"], "value": clone_json(payload)}
    if op == "delete":
        if not isinstance(value.get("key"), str):
            _invalid("delete requires a string key")
        return {"op": "delete", "key": value["key"]}
    if op == "put":
        item = value.get("item")
        collection = value.get("collection")
        if (
            not isinstance(collection, str)
            or len(collection) == 0
            or not _is_plain_object(item)
            or not isinstance(item.get("id"), str)
        ):
            _invalid("put requires a collection and an item with a string id")
        _json_safe(value["item"], "put item")
        return {"op": "put", "collection": value["collection"], "item": clone_json(value["item"])}
    if op == "remove":
        collection = value.get("collection")
        if (
            not isinstance(collection, str)
            or len(collection) == 0
            or not isinstance(value.get("id"), str)
        ):
            _invalid("remove requires string collection and id")
        return {"op": "remove", "collection": value["collection"], "id": value["id"]}
    raise AtomicMutationRejectedError(
        "unsupported-operation", f"unsupported atomic operation: {op}"
    )


def _normalize_condition(value: Any) -> dict[str, Any]:
    if not _is_plain_object(value):
        _invalid("invalid atomic condition")
    target = _target_from_unknown(value.get("target"))
    expected = value.get("expected")
    if expected in ("missing", "present"):
        return {"target": target, "expected": expected}
    if expected == "version" and isinstance(value.get("version"), str):
        return {"target": target, "expected": "version", "version": value["version"]}
    return _invalid("invalid atomic condition expectation")


@dataclass(frozen=True, slots=True)
class ValidatedRequest:
    """A canonicalized request plus its digest. Nothing has been written yet."""

    conditions: list[dict[str, Any]]
    operations: list[dict[str, Any]]
    idempotency: dict[str, Any] | None
    outcome: Any
    request_digest: str

    @property
    def has_outcome(self) -> bool:
        return not isinstance(self.outcome, _Undefined)

    @property
    def idempotency_key(self) -> str | None:
        return None if self.idempotency is None else str(self.idempotency["key"])


def _byte_length(text: str) -> int:
    return len(text.encode("utf-8"))


def validate_atomic_request(
    request: Any,
    limits: AtomicMutationLimits = DEFAULT_ATOMIC_LIMITS,
) -> ValidatedRequest:
    """Validate and canonicalize a request before any backend decision point."""
    if not _is_plain_object(request):
        _invalid("request must be a plain object")
    raw_operations: list[Any] = cast("list[Any]", request.get("operations"))
    if not isinstance(request.get("operations"), list) or len(raw_operations) == 0:
        _invalid("operations must be a non-empty array")
    if len(raw_operations) > limits.maxOperations:
        raise AtomicMutationRejectedError(
            "operation-limit-exceeded",
            f"request has {len(raw_operations)} operations;"
            f" this store's maxOperations is {limits.maxOperations}",
        )
    raw_conditions = request.get("conditions")
    if "conditions" in request and not isinstance(raw_conditions, list):
        _invalid("conditions must be an array")
    condition_source: list[Any] = (
        cast("list[Any]", raw_conditions) if isinstance(raw_conditions, list) else []
    )
    if len(condition_source) > limits.maxConditions:
        raise AtomicMutationRejectedError(
            "condition-limit-exceeded",
            f"request has {len(condition_source)} conditions;"
            f" this store's maxConditions is {limits.maxConditions}",
        )

    conditions = [_normalize_condition(entry) for entry in condition_source]
    condition_keys: set[str] = set()
    for condition in conditions:
        key = target_key(condition["target"])
        if key in condition_keys:
            _invalid("repeated conditions for one target are not supported")
        condition_keys.add(key)
    conditions.sort(key=cmp_to_key(lambda a, b: compare_targets(a["target"], b["target"])))

    operations = [_normalize_operation(entry) for entry in raw_operations]
    operation_keys: set[str] = set()
    for operation in operations:
        key = target_key(operation_target(operation))
        if key in operation_keys:
            _invalid("repeated operation targets are not supported")
        operation_keys.add(key)

    idempotency: dict[str, Any] | None = None
    outcome: Any = UNDEFINED
    if "idempotency" in request and request["idempotency"] is not None:
        raw = request["idempotency"]
        if not _is_plain_object(raw) or not isinstance(raw.get("key"), str):
            _invalid("idempotency requires a string key")
        idempotency = {"key": raw["key"]}
        if "outcome" in raw:
            _json_safe(raw["outcome"], "idempotency outcome")
            outcome = clone_json(raw["outcome"])
            outcome_bytes = _byte_length(canonical_json(outcome))
            if outcome_bytes > MAX_ATOMIC_OUTCOME_BYTES:
                raise AtomicMutationRejectedError(
                    "outcome-size-exceeded",
                    f"outcome is {outcome_bytes} bytes;"
                    f" the fixed outcome cap is {MAX_ATOMIC_OUTCOME_BYTES} bytes",
                )
            idempotency["outcome"] = outcome
    elif "idempotency" in request:
        _invalid("idempotency requires a string key")

    digest_input: dict[str, Any] = {
        "schema": REQUEST_SCHEMA,
        "conditions": conditions,
        "operations": operations,
    }
    if not isinstance(outcome, _Undefined):
        digest_input["outcome"] = outcome
    encoded = canonical_json(digest_input)
    # The idempotency key is deliberately excluded from the digest (the same
    # mutation has the same request identity under any key), but it is still
    # part of the bounded wire request and must not provide an unbounded escape
    # from maxRequestBytes.
    request_encoding = (
        encoded
        if idempotency is None
        else canonical_json({**digest_input, "idempotency": {"key": idempotency["key"]}})
    )
    request_bytes = _byte_length(request_encoding)
    if request_bytes > limits.maxRequestBytes:
        raise AtomicMutationRejectedError(
            "request-size-exceeded",
            f"request is {request_bytes} bytes;"
            f" this store's maxRequestBytes is {limits.maxRequestBytes}",
        )
    return ValidatedRequest(
        conditions=conditions,
        operations=operations,
        idempotency=idempotency,
        outcome=outcome,
        request_digest=sha256_hex(encoded),
    )
